text/0000-stabilize-os-str-conversions.md

```diff
@@ -0,0 +1,139 @@
+  names are encoded as an array of bytes. For example tarballs do this as well
+  as scp. This also requires viewing OS strings as an array of bytes and
+  sometimes constructing an OS string from an array of bytes.
+* Many C libraries are not written with the same OS string abstraction that the
+  standard library has, so they ubiquitously use `char*` for paths. This means
+  that FFI bindings in Rust wanting to use `Path` instead must convert a list of
+  bytes to an from an OS string.
+* On Windows most robust APIs take a wide string (e.g. `&[u16]`) and for any of
+  the situations above producers need to transform a `&[u8]` path into a wide
+  string somehow.
+
+The crux of these scenarios is that an OS string needs to either be converted to
+`&[u8]`/`&[u16]` or it needs to be constructed from these contents. If code is
+mostly written for one platform then there normally isn't a problem. For example
+on Unix OS strings are freely convertible between byte arrays and back, and on
+Windows the same is true for `u16` arrays. Problems can arise, however, when a
+library wants to perform these operations across all platforms.
+
+The functions being stabilized in this RFC are intended to provide convenient,
+yet fallible helper methods for performing these conversions. Currently
+byte-oriented, on Unix the methods simply pass through all contents (as
+everything is bytes) and on Windows crossing the `u16` to `u8` boundary involves
+interpreting the contents as valid unicode. The methods are all fallible as the
+unicode interpretation on Windows may fail.
+
+These convenience functions enable crates to avoid brittle `#[cfg]` logic while
+supporting a large number of cases right off the bat for both Unix and Windows.
+
+# Detailed design
+
+First, functions for converting from an array of bytes into an OS string will be
+stabilized, the two signatures in question are:
+
+```rust
+impl OsString {
+    fn from_platform_bytes(bytes: Vec<u8>) -> Option<OsString>;
+}
+
+impl OsStr {
+    fn from_platform_bytes(bytes: &[u8]) -> Option<&OsStr>;
+}
+```
+
+> Note: the `OsString::from_bytes` function today has been renamed here and the
+> generics have been removed.
+
+The name contains "platform" to indicate that the function has platform specific
+behavior, and the semantics will be:
+
+* On Unix, return `Some(mem::transmute(bytes))`
+* On Windows, interpret the bytes as utf-8, returning `Some` if this succeeded
+  or `None` if it failed.
+
+Next, some convenience functions will be stabilized for extracting a list of
+bytes out of an OS string:
+
+```rust
+impl OsStr {
+    fn to_bytes(&self) -> Option<&[u8]>;
+    fn to_cstring(&self) -> Option<CString>;
+}
+```
+
+The semantics of these functions will be:
+
+* On Unix, return `Some` always on `to_bytes` and call `CString::new(..).ok()`
+  for `to_cstring`.
+* On Windows, attempt to interpret the `&[u16]` as utf-16, and if successful
```

Comment:
Implementation-wise, yes, but conceptually `OsString` on Windows is `&[u16]` so this is more a description of what's metaphorically happening rather than literally.